hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/ReconcileSubcommand.java

```diff
@@ -34,15 +48,188 @@
+    Iterator<String> idIterator;
+    // PicoCLI arity check guarantees at least one element will be present.
+    if (containerList.get(0).equals("-")) {
+      // Read from stdin.
+      idIterator = new Scanner(System.in, StandardCharsets.UTF_8.name());
+    } else {
+      // A list of containers was provided.
+      idIterator = containerList.iterator();
+    }
+
+    if (status) {
+      // Automatically creates one array for the output, while allowing us to flush each object individually.
+      try (SequenceWriter arrayWriter = JsonUtils.getSequenceWriter(System.out)) {
+        // Since status is retrieved using container info, do client side validation that it is only used for Ratis
+        // containers. If EC containers are given, print a  message to stderr and eventually exit non-zero, but continue
+        // processing the remaining containers.
+        int invalidCount = 0;
+        while (idIterator.hasNext()) {
+          long containerID = Long.parseLong(idIterator.next());
+          if (!printReconciliationStatus(scmClient, containerID, arrayWriter)) {
+            invalidCount++;
+          }
+        }
+        if (invalidCount > 0) {
+          throw new RuntimeException("Failed to process reconciliation status for " + invalidCount + " containers");
+        }
+      }
+    } else {
+      int invalidCount = 0;
+      while (idIterator.hasNext()) {
+        long containerID = Long.parseLong(idIterator.next());
+        try {
+          executeReconciliation(scmClient, containerID);
+        } catch (Exception ex) {
+          System.err.println("Failed to send reconciliation to container " + containerID + ": " + ex.getMessage());
+          invalidCount++;
+        }
+      }
+      if (invalidCount > 0) {
+        throw new RuntimeException("Failed trigger reconciliation for " + invalidCount + " containers");
+      }
+    }
+  }
+
+  private boolean printReconciliationStatus(ScmClient scmClient, long containerID, SequenceWriter arrayWriter)
+      throws IOException {
+    ContainerInfo containerInfo = scmClient.getContainer(containerID);
+    if (containerInfo.getReplicationType() != HddsProtos.ReplicationType.RATIS) {
+      System.err.println("Cannot get status of container " + containerID +
+          ". Reconciliation is only supported for Ratis replicated containers");
+      return false;
+    }
+    List<ContainerReplicaInfo> replicas = scmClient.getContainerReplicas(containerID);
+    arrayWriter.write(new ContainerWrapper(containerInfo, replicas));
+    arrayWriter.flush();
+    return true;
+  }
+
+  private void executeReconciliation(ScmClient scmClient, long containerID) throws IOException {
+    scmClient.reconcileContainer(containerID);
+    System.out.println("Reconciliation has been triggered for container " + containerID);
+    System.out.println("Use \"ozone admin container reconcile --status " + containerID + "\" to see the checksums of " +
+        "each container replica");
```

Comment:
We don't have a good way to indicate when reconciliation is completed... Maybe we add a reconciliation count as a metric to the container info?

Reply:
Agreed, some sort of feedback that it finished would be good. I was planning to revisit this with the SCM integration. For example, SCM may end up tracking the known in-flight reconciliations and could report the results. 

Are you suggesting adding a jmx metric or just a counter in the code? I'm not sure exactly how this work, but I imagine the counter would get persisted on the DN side and SCM would see the updates. Users would then look for the count to increase when they send in a command.